### PR TITLE
overlord/state: implement Task.Logf and add Task.Log/Errorf.

### PR DESCRIPTION
**Pull request by niemeyer**

### CODE REVIEW

File: overlord/state/task.go [hunk 1]
```diff
@@ -163,13 +167,48 @@ func (t *Task) SetProgress(cur, total int) {
...
+const (
+	// Messages logged in tasks are guaranteed to use the following strings
+	// plus ": " as a prefix, so these may be handled programatically and
+	// stripped for presentation.
+	LogInfo  = "INFO"
+	LogError = "ERROR"
+)
+
+func (t *Task) logf(kind, format string, args ...interface{}) {
+	if len(t.log) > 9 {
```
Comment (zyga) [inline overlord/state/task.go]: So we-log rotate at 10 items? I don't really mind it here but there are more efficient ways to keep a circular log.

**Comment by zyga** [comment]
Just one comment, plus please merge the branch to resolve conflicts.

File: overlord/state/task_test.go [hunk 2]
```diff
@@ -229,33 +168,81 @@ func (ts *taskSuite) TestTaskWaitFor(c *C) {
...
+func (ts *taskSuite) TestNeedsLock(c *C) {
+	st := state.New(nil)
+	st.Lock()
+	chg := st.NewChange("install", "...")
+	t1 := chg.NewTask("download", "1...")
+	t2 := chg.NewTask("install", "2...")
 	st.Unlock()
 
-	c.Assert(func() { t.HaltTasks() }, PanicMatches, "internal error: accessing state without lock")
+	funcs := []func() {
```
Comment (mvo5) [inline overlord/state/task_test.go]: Nice!

**Comment by mvo5** [comment]
LTGTM

File: overlord/state/task.go (hunk 1 above)
Comment (niemeyer) [inline overlord/state/task.go]: Yes, but I doubt they'll be simpler, and copying 10 pointers when logging something is nothing compared to running Sprintf. Let's please make this work and learn about the real usage patterns we will develop before investing further.

**Comment by niemeyer** [comment]
Conflicts fixed.

**Comment by pedronis** [comment]
seems to need a "go fmt" pass in overlord/task

**Comment by pedronis** [comment]
Also I'm a bit uncomfortable with the fact that a few of Logf can hide a previous Errorf, though at least the status will be correct

**Comment by niemeyer** [comment]
Yeah, I thought about that as well.. We can be smarter about the way we handle this, but I expect the pattern in general to be either error is last, or it means there's activity still going, which means the error is not so relevant. Let's get started and observe how we use it.

**Pull request merged**